File: app/store.py
"""
In-memory store for shortened URLs and their click analytics.
 
Two concurrency points worth noting:
 
1. Click counting uses an asyncio.Lock per URL entry rather than one global
   lock, so a burst of clicks on one URL doesn't block reads/writes to any
   other URL. This is the same per-key locking pattern used in the rate
   limiter's TokenBucket.
 
2. Click events are recorded *after* the redirect response is sent
   (via FastAPI BackgroundTasks), so the hot-path redirect latency is
   never affected by analytics bookkeeping. The store is designed for
   that: record_click() is async and safe to call from a background task.
"""

from __future__ import annotations
import asyncio
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional
from urllib.parse import urlparse

from app.models import ClickEvent

@dataclass
class URLEntry:
    code: str
    original_url: str
    created_at: datetime
    expires_at: Optional[datetime]
    max_clicks: int

    # Protected by _lock
    click_count: int = 0
    clicks: list[ClickEvent] = field(default_factory=list)
    referrer_counts: dict[str, int] = field(default_factory=lambda: defaultdict(int))
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock, repr=False)

    def is_expired(self) -> bool:
        if self.expires_at is None:
            return False
        return datetime.now(timezone.utc) > self.expires_at
    
    async def record_click(self, referrer: str | None, user_agent: str | None) -> None:
        async with self._lock:
            self.click_count += 1

            event = ClickEvent(
                timestamp=datetime.now(timezone.utc),
                referrer=referrer,
                user_agent=user_agent,
            )
            self.clicks.append(event)
            # Ring-buffer: drop oldest clicks beyond the cap
            if len(self.clicks) > self.max_clicks:
                self.clicks = self.clicks[-self.max_clicks:]

            if referrer:
                domain = _extract_domain(referrer)
                self.referrer_counts[domain] += 1

    async def get_stats(self) -> dict:
        async with self._lock:
            return {
                "click_count": self.click_count,
                "referrers": dict(self.referrer_counts),
                "recent_clicks": list(reversed(self.clicks)),
            }
        

def _extract_domain(referrer: str) -> str:
    try:
        parsed = urlparse(referrer)
        return parsed.netloc or referrer
    except Exception:
        return referrer


class URLStore:
    def __init__(self) -> None:
        self._entries: dict[str, URLEntry] = {}
        self._lock = asyncio.Lock()

    async def save(self, entry: URLEntry) -> None:
        async with self._lock:
            self._entries[entry.code] = entry

    async def get(self, code: str) -> URLEntry | None:
        """Returns the entry if it exists and hasn't expired."""
        async with self._lock:
            entry = self._entries.get(code)
        if entry is None:
            return None
        if entry.is_expired():
            async with self._lock:
                self._entries.pop(code, None)
            return None
        return entry
    
    async def exists(self, code: str) -> bool:
        async with self._lock:
            return code in self._entries
        
    async def delete(self, code: str) -> bool:
        async with self._lock:
            self._entries.pop(code, None) is not None

    async def evict_expired(self) -> int:
        now = datetime.now(timezone.utc)
        async with self._lock:
            stale = [
                code for code, entry in self._entries.items()
                if entry.expires_at is not None and now > entry.expires_at
            ]
            for code in stale:
                del self._entries[code]
        return len(stale)
    
    def __len__(self) -> int:
        return len(self._entries)

File: internal/store/store.go
// Holds shortened URLs and their click analytics in memory.

package store

import (
	"context"
	"net/url"
	"sync"
	"time"

	"url-shortener/internal/models"
)

// URLEntry holds one shortened URL and all its analytics state.
type URLEntry struct {
	Code        string
	OriginalURL string
	CreatedAt   time.Time
	ExpiresAt   *time.Time
	MaxClicks   int

	// Everything below is protected by mu.
	mu            sync.Mutex
	ClickCount    int64
	Clicks        []models.ClickEvent
	ReferrerCount map[string]int64
}

func NewEntry(code, originalURL string, createdAt time.Time, expiresAt *time.Time, maxClicks int) *URLEntry {
	return &URLEntry{
		Code:          code,
		OriginalURL:   originalURL,
		CreatedAt:     createdAt,
		ExpiresAt:     expiresAt,
		MaxClicks:     maxClicks,
		ReferrerCount: make(map[string]int64),
	}
}

func (e *URLEntry) IsExpired() bool {
	if e.ExpiresAt == nil {
		return false
	}
	return time.Now().After(*e.ExpiresAt)
}

// RecordClick is safe to call from a goroutine after the response is sent.
// It increments the counter, appends the event, and trims the ring buffer.
func (e *URLEntry) RecordClick(referrer, userAgent string) {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.ClickCount++

	event := models.ClickEvent{
		Timestamp: time.Now().UTC(),
		Referrer:  referrer,
		UserAgent: userAgent,
	}
	e.Clicks = append(e.Clicks, event)

	// Ring buffer: keep only the most recent MaxClicks events.
	if len(e.Clicks) > e.MaxClicks {
		e.Clicks = e.Clicks[len(e.Clicks)-e.MaxClicks:]
	}

	if referrer != "" {
		domain := extractDomain(referrer)
		e.ReferrerCount[domain]++
	}
}

// Stats returns a snapshot of click analytics under the lock.
func (e *URLEntry) Stats() (int64, map[string]int64, []models.ClickEvent) {
	e.mu.Lock()
	defer e.mu.Unlock()

	refs := make(map[string]int64, len(e.ReferrerCount))
	for k, v := range e.ReferrerCount {
		refs[k] = v
	}

	// Return newest-first, same as the Python version.
	clicks := make([]models.ClickEvent, len(e.Clicks))
	for i, c := range e.Clicks {
		clicks[len(e.Clicks)-1-i] = c
	}

	return e.ClickCount, refs, clicks
}

func extractDomain(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil || u.Host == "" {
		return rawURL
	}
	return u.Host
}

// URLStore is the map of all live entries, protected by an RWMutex.
// Read-heavy operations (Get, Exists) take the read lock; writes take the full lock.
type URLStore struct {
	mu      sync.RWMutex
	entries map[string]*URLEntry
}

func NewStore() *URLStore {
	return &URLStore{entries: make(map[string]*URLEntry)}
}

func (s *URLStore) Save(entry *URLEntry) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.entries[entry.Code] = entry
}

func (s *URLStore) Get(code string) (*URLEntry, bool) {
	s.mu.RLock()
	entry, ok := s.entries[code]
	s.mu.RUnlock()

	if !ok {
		return nil, false
	}

	if entry.IsExpired() {
		s.mu.Lock()
		delete(s.entries, code)
		s.mu.Unlock()
		return nil, false
	}

	return entry, true
}

func (s *URLStore) Exists(code string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	_, ok := s.entries[code]

	return ok
}

func (s *URLStore) Delete(code string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.entries[code]; !ok {
		return false
	}

	delete(s.entries, code)

	return true
}

func (s *URLStore) EvictExpired() int {
	now := time.Now()
	s.mu.Lock()
	defer s.mu.Unlock()

	n := 0
	for code, entry := range s.entries {
		if entry.ExpiresAt != nil && now.After(*entry.ExpiresAt) {
			delete(s.entries, code)
			n++
		}
	}

	return n
}

func (s *URLStore) Len() int {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return len(s.entries)
}

// Background goroutine that periodically evicts expired entries. Stops cleanly when ctx is cancelled.
func (s *URLStore) StartCleanup(ctx context.Context, interval time.Duration) {
	go func() {
		t := time.NewTicker(interval)
		defer t.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-t.C:
				s.EvictExpired()
			}
		}
	}()
}